irr/src/CGLTFMeshFileLoader.cpp

```diff
@@ -0,0 +1,610 @@
+				if (index == std::numeric_limits<u8>::max())
+					throw std::runtime_error("invalid index");
+				break;
+			}
+			case tiniergltf::Accessor::ComponentType::UNSIGNED_SHORT: {
+				index = readPrimitive<u16>(BufferOffset(buf, elemIdx * sizeof(u16)));
+				if (index == std::numeric_limits<u16>::max())
+					throw std::runtime_error("invalid index");
+				break;
+			}
+			case tiniergltf::Accessor::ComponentType::UNSIGNED_INT: {
+				u32 indexWide = readPrimitive<u32>(BufferOffset(buf, elemIdx * sizeof(u32)));
+				// Use >= here for consistency.
+				if (indexWide >= std::numeric_limits<u16>::max())
+					throw std::runtime_error("index too large (>= 65536)");
+				index = static_cast<u16>(indexWide);
+				break;
+			}
+			default:
+				throw std::runtime_error("invalid index component type");
+		}
+		indices.push_back(index);
+	}
+
+	return indices;
+}
+
+/**
+ * Create a vector of video::S3DVertex (model data) from a mesh & primitive index.
+*/
+std::optional<std::vector<video::S3DVertex>> CGLTFMeshFileLoader::MeshExtractor::getVertices(
+		const std::size_t meshIdx,
+		const std::size_t primitiveIdx) const
+{
+	const auto positionAccessorIdx = getPositionAccessorIdx(
+			meshIdx, primitiveIdx);
+	if (!positionAccessorIdx.has_value()) {
+		// "When positions are not specified, client implementations SHOULD skip primitive's rendering"
+		return std::nullopt;
+	}
+
+	std::vector<vertex_t> vertices{};
+	vertices.resize(getElemCount(*positionAccessorIdx));
+	copyPositions(*positionAccessorIdx, vertices);
+
+	const auto normalAccessorIdx = getNormalAccessorIdx(
+			meshIdx, primitiveIdx);
+	if (normalAccessorIdx.has_value()) {
+		copyNormals(normalAccessorIdx.value(), vertices);
+	}
+
+	const auto tCoordAccessorIdx = getTCoordAccessorIdx(
+			meshIdx, primitiveIdx);
+	if (tCoordAccessorIdx.has_value()) {
+		copyTCoords(tCoordAccessorIdx.value(), vertices);
+	}
+
+	return vertices;
+}
+
+/**
+ * Get the amount of meshes that a model contains.
+*/
+std::size_t CGLTFMeshFileLoader::MeshExtractor::getMeshCount() const
+{
+	return m_gltf_model.meshes->size();
+}
+
+/**
+ * Get the amount of primitives that a mesh in a model contains.
+*/
+std::size_t CGLTFMeshFileLoader::MeshExtractor::getPrimitiveCount(
+		const std::size_t meshIdx) const
+{
+	return m_gltf_model.meshes->at(meshIdx).primitives.size();
+}
+
+/**
+ * Templated buffer reader. Based on type width.
+ * This is specifically used to build upon to read more complex data types.
+ * It is also used raw to read arrays directly.
+ * Basically we're using the width of the type to infer 
+ * how big of a gap we have from the beginning of the buffer.
+*/
+template <typename T>
+T CGLTFMeshFileLoader::MeshExtractor::readPrimitive(
+		const BufferOffset& readFrom)
+{
+	unsigned char d[sizeof(T)]{};
+	for (std::size_t i = 0; i < sizeof(T); ++i) {
+		d[i] = readFrom.at(i);
+	}
+	T dest;
+	std::memcpy(&dest, d, sizeof(dest));
+	return dest;
```

Comment:
missing endian handling

Reply:
done in https://github.com/minetest/minetest/pull/14557/commits/eea01c3957308c7bc1d0a8fe03dd2324447d778d